Tests/NIOTests/SyscallAbstractionLayer.swift

```diff
@@ -0,0 +1,665 @@
+            try self.assertRegister { selectable, eventSet, registration in
+                if case .socketChannel(let channel, let registrationEventSet) = registration {
+                    XCTAssertEqual(localAddress, channel.localAddress)
+                    XCTAssertEqual(remoteAddress, channel.remoteAddress)
+                    XCTAssertEqual(eventSet, registrationEventSet)
+                    XCTAssertEqual(.reset, eventSet)
+                    return true
+                } else {
+                    return false
+                }
+            }
+            try self.assertReregister { selectable, eventSet in
+                XCTAssertEqual([.reset, .readEOF], eventSet)
+                return true
+            }
+            // because autoRead is on by default
+            try self.assertReregister { selectable, eventSet in
+                XCTAssertEqual([.reset, .readEOF, .read], eventSet)
+                return true
+            }
+        }) {
+            channel.register().flatMap {
+                channel.connect(to: remoteAddress)
+            }
+        }
+        XCTAssertNoThrow(try connectFuture.teleportWait())
+        return channel
+    }
+
+    func tearDownSAL() {
+        SAL.printIfDebug("=== TEAR DOWN ===")
+        XCTAssertNotNil(self.kernelToUserBox)
+        XCTAssertNotNil(self.userToKernelBox)
+        XCTAssertNotNil(self.wakeups)
+        XCTAssertNotNil(self.group)
+
+        let group = DispatchGroup()
+        group.enter()
+        XCTAssertNoThrow(self.group.shutdownGracefully(queue: DispatchQueue.global()) { error in
+            XCTAssertNil(error, "unexpected error: \(error!)")
+            group.leave()
+        })
+        XCTAssertNoThrow(try self.assertWakeup())
+        group.wait()
+
+        self.group = nil
+        self.kernelToUserBox = nil
+        self.userToKernelBox = nil
+        self.wakeups = nil
+    }
+
+    func assertParkedRightNow(file: StaticString = #file, line: UInt = #line) throws {
+        SAL.printIfDebug("\(#function)")
+        let syscall = try self.userToKernelBox.waitForValue()
+        if case .whenReady = syscall {
+            return
+        } else {
+            XCTFail("unexpected syscall \(syscall)", file: file, line: line)
+        }
+    }
+
+    func assertWaitingForNotification(result: SelectorEvent<NIORegistration>?,
+                                      file: StaticString = #file, line: UInt = #line) throws {
+        SAL.printIfDebug("\(#function)")
+        try self.selector.assertSyscallAndReturn(.returnSelectorEvent(result),
+                                                 file: file, line: line) { syscall in
+            if case .whenReady = syscall {
+                return true
+            } else {
+                return false
+            }
+        }
+    }
+
+    func assertWakeup(file: StaticString = #file, line: UInt = #line) throws {
+        try self.selector.assertWakeup(file: file, line: line)
```

Comment:
`printIfDebug`?

Reply:
that's done by `self.selector.assertWakeup`